alacritty/src/display/mod.rs

```diff
@@ -980,74 +1020,203 @@ impl Display {
     }
 
     /// Format search regex to account for the cursor and fullwidth characters.
-    fn format_search(size_info: &SizeInfo, search_regex: &str, search_label: &str) -> String {
-        // Add spacers for wide chars.
-        let mut formatted_regex = String::with_capacity(search_regex.len());
-        for c in search_regex.chars() {
-            formatted_regex.push(c);
-            if c.width() == Some(2) {
-                formatted_regex.push(' ');
+    fn format_search(search_regex: &str, search_label: &str, max_width: usize) -> String {
+        let label_len = search_label.len();
+
+        // If the label doesn't fit into screen return.
+        if label_len > max_width {
+            return search_label[..max_width].to_owned();
+        }
+
+        // The search string consists of `search_label` + `search_regex` + `cursor`.
+        let mut bar_text = String::from(search_label);
+        bar_text.extend(
+            Self::format_visible_text(
+                iter::once('_').chain(search_regex.chars().rev()),
+                max_width - label_len,
+                None,
+            )
+            .into_iter()
+            .rev(),
+        );
+
+        bar_text
+    }
+
+    /// Format text to fit into visible on-screen element, like `footer_bar`.
+    pub fn format_visible_text(
+        chars: impl Iterator<Item = char>,
+        max_width: usize,
+        shortener: Option<char>,
+    ) -> Vec<char> {
+        let mut visible_text = Vec::<char>::with_capacity(max_width);
+        let mut num_chars = 0;
+
+        for ch in chars {
+            let char_width = ch.width().unwrap_or(1);
+
+            // If the text doesn't fit into `max_width` terminate iteration.
+            if num_chars + char_width > max_width {
+                // Put shortener if it was provided to indicate that text got shortened.
+                let shortener = match shortener {
+                    Some(shortener) => shortener,
+                    None => break,
+                };
+
+                let shortener_position = if num_chars < max_width {
+                    // Allocate space for the char we're about to write.
+                    visible_text.push(' ');
+                    num_chars
+                } else if num_chars > 1
+                    && visible_text.get(num_chars - 2).and_then(|ch| ch.width()) == Some(2)
+                {
+                    // Remove the trailing space.
+                    let _ = visible_text.pop();
+                    num_chars - 2
+                } else {
+                    num_chars - 1
+                };
+
+                visible_text[shortener_position] = shortener;
+
+                break;
             }
+
+            visible_text.push(ch);
+
+            // Add spacers for wide chars.
+            if char_width == 2 {
+                visible_text.push(' ');
+            }
+
+            num_chars += char_width;
```

Comment:
I recommend putting this in an `else` block rather than `break`ing, it's way too big of a conditional block with the early return to put an early return just to skip 5 lines.